### PR TITLE
Implement unstructured cassandra and dse yaml configuration

### LINKED ISSUE #652: K8SSAND-1733 ⁃ Remove PreserveUnknownFields from K8ssandraCluster CRD
**What is missing?**

The `CassandraDatacenterTemplate` type is declared with the kubebuilder annotation `//+kubebuilder:pruning:PreserveUnknownFields`, which leads to lighter touch OpenAPI validation on the type. This should be removed if possible.

**Why do we need it?**

Invalid fields can be added to a CR without triggering an error, which leads to a poor user experience when those fields then cause silent reconciliation failures.

See [here](https://github.com/k8ssandra/k8ssandra-operator/issues/647#issuecomment-1210890425) for an example.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1733) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1733
┆priority: Medium

### LINKED ISSUE #710: K8SSAND-1819 ⁃ Implement unstructured cassandra and dse yaml configuration
**What is missing?**

A dynamic way of passing unstructured options to cassandra.yaml and dse.yaml in a K8ssandraCluster object.

**Why do we need it?**

Because maintaining structured YAML options is too time-consuming.

**Environment**

* K8ssandra Operator version:

    `Insert image tag or Git SHA here`
    
    **Anything else we need to know?**:

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1819) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1819
┆priority: Medium

### PR DESCRIPTION
**Pull request by adutra**
**What this PR does**:

Introduces unstructured data for cassandra.yaml and dse.yaml.

**Which issue(s) this PR fixes**:
Fixes #710 
Fixes #652 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: config/crd/kustomization.yaml [hunk 1]
```diff
@@ -16,14 +16,6 @@ resources:
 - bases/medusa.k8ssandra.io_medusabackupschedules.yaml
 #+kubebuilder:scaffold:crdkustomizeresource
 
-patches:
```
Comment (adutra) [inline config/crd/kustomization.yaml:19]: This fixes #652: if you remove the `+kubebuilder:pruning:PreserveUnknownFields` annotation from `CassandraDatacenterTemplate`, this patch gets applied and removes everything in the config.

File: controllers/k8ssandra/k8ssandracluster_controller_test.go [hunk 2]
```diff
@@ -400,13 +395,6 @@ func applyClusterTemplateConfigs(t *testing.T, ctx context.Context, f *framework
 	assert.Equal(dc1Size, dc1.Spec.Size)
 	assert.Equal(dc1.Spec.SuperuserSecretName, superUserSecretName)
 
-	actualConfig, err := gabs.ParseJSON(dc1.Spec.Config)
```
Comment (adutra) [inline controllers/k8ssandra/k8ssandracluster_controller_test.go:403]: I removed this kind of test code. All it does is serialize the config twice, and assert that the first result is equal to the second.

File: controllers/k8ssandra/k8ssandracluster_controller_test.go [hunk 3]
```diff
@@ -747,46 +707,13 @@ func applyClusterTemplateAndDatacenterTemplateConfigs(t *testing.T, ctx context.
...
-	assert.Equal(expectedConfig, actualConfig)
-
 	t.Log("deleting K8ssandraCluster")
 	err = f.DeleteK8ssandraCluster(ctx, client.ObjectKey{Namespace: kc.Namespace, Name: kc.Name}, timeout, interval)
 	require.NoError(err, "failed to delete K8ssandraCluster")
 	f.AssertObjectDoesNotExist(ctx, t, dc1Key, &cassdcapi.CassandraDatacenter{}, timeout, interval)
 	f.AssertObjectDoesNotExist(ctx, t, dc2Key, &cassdcapi.CassandraDatacenter{}, timeout, interval)
 }
 
-func parseCassandraConfig(config *api.CassandraConfig, serverVersion string, systemRF int, dcNames ...string) (*gabs.Container, error) {
```
Comment (adutra) [inline controllers/k8ssandra/k8ssandracluster_controller_test.go:764]: This is the function that was being used to test serialization of config by mimicking the behavior of the actual serialization function.

File: pkg/cassandra/config.go [hunk 4]
```diff
@@ -3,153 +3,99 @@ package cassandra
...
+
+	out := make(unstructured.Unstructured)
+
+	// cassandra.yaml is an unstructured map, we simply append it to the output
+	// TODO validate cassandra.yaml settings with JSON schemas
+	// TODO postprocess cassandra.yaml settings, e.g. mountable volumes or resource.Quantity syntax
+	if len(config.CassandraYaml) > 0 {
+		out["cassandra-yaml"] = config.CassandraYaml
 	}
-	addNumTokens(apiConfig, cfg)
```
Comment (adutra) [inline pkg/cassandra/config.go:28]: I moved all mutating functions outside of this function; I would like this function to be devoted solely to producing the final JSON by marshaling the input config.

File: pkg/cassandra/config.go [hunk 5]
```diff
@@ -3,153 +3,99 @@ package cassandra
...
 	// Even though we default to Cassandra's stock defaults for num_tokens, we need to
 	// explicitly set it because the config builder defaults to num_tokens: 1
-	if cfg.CassandraYaml.NumTokens == nil {
-		version := template.ServerVersion
-		if template.ServerType == api.ServerDistributionCassandra && version.Major() == 3 {
-			cfg.CassandraYaml.NumTokens = pointer.Int(256)
-		} else {
-			cfg.CassandraYaml.NumTokens = pointer.Int(16)
-		}
+	// FIXME when importing existing clusters, we should not override the user's num_tokens setting – even if it's unset (1)
```
Comment (adutra) [inline pkg/cassandra/config.go:54]: This FIXME has been suggested to me by @burmanm: when importing existing clusters, we shouldn't add `num_tokens` blindly.

File: pkg/cassandra/config.go [hunk 6]
```diff
@@ -3,153 +3,99 @@ package cassandra
...
+	if template.ServerType == api.ServerDistributionCassandra && template.ServerVersion.Major() == 3 {
+		template.CassandraConfig.CassandraYaml.PutIfAbsent("num_tokens", 256)
+	} else {
+		template.CassandraConfig.CassandraYaml.PutIfAbsent("num_tokens", 16)
 	}
 }
 
-func addEncryptionOptions(template *DatacenterConfig, cfg *cassConfig) {
```
Comment (adutra) [inline pkg/cassandra/config.go:67]: Moved these to another existing function called `handleEncryptionOptions` declared in encryption.go. I think it's easier if these 2 functions are merged together.

File: pkg/cassandra/config.go [hunk 7]
```diff
@@ -3,153 +3,99 @@ package cassandra
...
-	}
-
-	if cassandraYaml.CommitlogSyncPeriodInMs != nil && cassandraYaml.CommitlogSyncBatchWindowInMs != nil {
-		return fmt.Errorf("CommitlogSyncPeriodInMs and CommitlogSyncBatchWindowInMs are mutually exclusive")
+// validateCassandraYaml provides semantic validation for cassandra.yaml settings.
+// TODO this is a relic of the structured YAML approach. Only the bits that are still relevant were ported over.
+// From now on, all syntactic validation is expected to happen externally using JSON schemas.
+// Do not use this function to validate the syntax of cassandra.yaml settings. Use ONLY to validate
+// any semantics that cannot be validated with JSON schemas.
+func validateCassandraYaml(cassandraYaml unstructured.Unstructured) error {
```
Comment (adutra) [inline pkg/cassandra/config.go:94]: Most of the validations done here are gone for now.

File: pkg/cassandra/config_test.go [hunk 8]
```diff
@@ -113,3702 +111,116 @@ func TestApplySystemReplication(t *testing.T) {
...
+		serverVersion   *semver.Version
+		serverType      api.ServerDistribution
+		cassandraConfig api.CassandraConfig
+		got             []byte
+		want            string
 	}
 
 	tests := []test{
 		{
-			name: "[4.0.0] concurrent_reads, concurrent_writes, concurrent_counter_writes",
```
Comment (adutra) [inline pkg/cassandra/config_test.go:130]: Most of these tests do no make sense anymore. I kept a few ones as smoke tests.

File: pkg/cassandra/datacenter.go [hunk 9]
```diff
@@ -145,7 +144,11 @@ func NewDatacenter(klusterKey types.NamespacedName, template *DatacenterConfig)
 		return nil, err
 	}
 
-	rawConfig, err := CreateJsonConfig(template)
+	handleDeprecatedJvmOptions(&template.CassandraConfig.JvmOptions)
```
Comment (adutra) [inline pkg/cassandra/datacenter.go:147]: Moved mutating functions here.

File: test/testdata/fixtures/multi-dc-encryption-medusa/k8ssandra.yaml [hunk 10]
```diff
@@ -54,7 +54,6 @@ spec:
         thrift_prepared_statements_cache_size_mb: 1
         prepared_statements_cache_size_mb: 1
         slow_query_log_timeout_in_ms: 0
-        block_for_peers_timeout_in_secs: 10000
```
Comment (adutra) [inline test/testdata/fixtures/multi-dc-encryption-medusa/k8ssandra.yaml:57]: Many options were invalid in our e2e tests. They were being filtered out so far, but not anymore.

File: test/testdata/fixtures/single-dc-dse/k8ssandra.yaml [hunk 11]
```diff
@@ -30,343 +30,11 @@ spec:
             authorizer: com.datastax.bdp.cassandra.auth.DseAuthorizer
             role_manager: com.datastax.bdp.cassandra.auth.DseRoleManager
           dseYaml:
-            advanced_replication_options:
```
Comment (adutra) [inline test/testdata/fixtures/single-dc-dse/k8ssandra.yaml:33]: It does not make sense anymore to test every possible option.

File: pkg/utils/pointers.go [hunk 12]
```diff
@@ -1,14 +0,0 @@
-package utils
-
-import "reflect"
-
-func IsNil(v interface{}) bool {
```
Comment (adutra) [inline pkg/utils/pointers.go:5]: Not used anymore.

File: pkg/utils/map.go [hunk 13]
```diff
@@ -34,7 +36,7 @@ func MergeMapNested(allowOverwrite bool, sources ...map[string]interface{}) (map
 						} else {
 							destination[k] = merged
 						}
-					} else if allowOverwrite {
+					} else if allowOverwrite || destVal == nil {
```
Comment (adutra) [inline pkg/utils/map.go:39]: This small modification in `MergeMapNested` and `PutMapNested` was motivated by the fact that existing map entries with nil values were treated like non-nil ones, which is not what we want when merging or updating configs.

File: pkg/utils/map_test.go [hunk 14]
```diff
@@ -55,6 +55,13 @@ func TestMergeMapNested(t *testing.T) {
 			nil,
 			assert.Error,
 		},
+		{
+			"simple merge overlapping no overwrite nil value",
```
Comment (adutra) [inline pkg/utils/map_test.go:59]: Added a few unit tests around nil values to make sure the changes in map.go are OK.

File: pkg/utils/map.go [hunk 15]
```diff
@@ -19,6 +19,8 @@ func MergeMap(sources ...map[string]string) map[string]string {
 // MergeMapNested will take two or more maps, merging the entries of the sources map into a
 // destination map. If both maps share the same key then destination's value for that key will be
 // merged with what's in source. The source maps are not modified.
+// If allowOverwrite is false, an error is returned if the key already exists and its value is not
+// nil.
 func MergeMapNested(allowOverwrite bool, sources ...map[string]interface{}) (map[string]interface{}, error) {
```
Comment (adejanovski) [inline pkg/utils/map.go:24]: Now that our config is unstructured, do you think we could use this to merge cluster level and dc level Cassandra configs? Some folks have had issues with this as they were defining some settings at cluster level and others at dc level, but then all the cluster level ones were deleted (we're not doing any merge).

**Review by adejanovski — APPROVED** [review]
This is some very solid work with a lot of great refactoring!
Nice to see 16k+ LoC go away 💪 
There was one failing e2e test. I've restarted the multi e2e test workflow, let's wait for all of them to pass before merging.

File: pkg/utils/map.go (hunk 15 above)
Comment (adutra) [reply pkg/utils/map.go:24]: That's a very good idea, so I created an issue: https://github.com/k8ssandra/k8ssandra-operator/issues/715

We must keep in mind though: JVM options are still structured, so we can't use this function for them.

**Pull request merged**